bindings/rust/s2n-tls/src/connection.rs

```diff
@@ -373,18 +373,55 @@ impl Connection {
     pub fn poll_negotiate(&mut self) -> Poll<Result<&mut Self, Error>> {
         let mut blocked = s2n_blocked_status::NOT_BLOCKED;
 
-        // check if an async task for the client_hello_callback exists and
-        // poll it to completion
-        if let Some(fut) = self.take_connection_future() {
-            if poll_client_hello_callback(self, Some(fut)).is_pending() {
-                return Poll::Pending;
+        loop {
+            // check if an async task exists and poll it to completion
+            if let Some(fut) = self.take_connection_future() {
+                match Self::poll_async_task(self, fut) {
+                    Poll::Ready(res) => {
+                        // Handle the result from the connection future
+                        // - happy case: continue to calling s2n_negotiate and make
+                        //               progress on the handshake
+                        // - error case: if the callback returned an error then abort
+                        //               the handshake
+                        if let Err(err) = res {
```

Comment:
Why not flatten the match here? That way your comment above will only apply to a single condition

Reply:
https://github.com/aws/s2n-tls/pull/3687/commits/4f2d6aa519b1f172ad4847bce7d6619e6b4e1637